backend/libraries/ballerina-core-python/tests/serialization_test.py

```diff
@@ -0,0 +1,258 @@
+        assert value == (42, "hello")
+
+    @staticmethod
+    def test_dict_serializer() -> None:
+        value = {"a": 42, "b": 43}
+        serializer = dict_serializer(str_serializer, int_serializer)
+        serialized = serializer(value)
+        assert serialized == {"a": 42, "b": 43}
+
+    @staticmethod
+    def test_dict_deserializer() -> None:
+        serialized: Json = {"a": 42, "b": 43}
+        deserializer = dict_deserializer(str_deserializer, int_deserializer)
+        value = deserializer(serialized)
+        assert value == {"a": 42, "b": 43}
+
+
+class TestOptionSerializer:
+    @staticmethod
+    def test_option_serializer_some() -> None:
+        value = Option.some(42)
+        serializer = option_serializer(int_serializer)
+        serialized = serializer(value)
+        assert serialized == {"discriminator": "some", "value": 42}
+
+    @staticmethod
+    def test_option_serializer_none() -> None:
+        value: Option[int] = Option.nothing()
+        serializer = option_serializer(int_serializer)
+        serialized = serializer(value)
+        assert serialized == {"discriminator": "nothing", "value": None}
+
+    @staticmethod
+    def test_option_deserializer_some() -> None:
+        serialized: Json = {"discriminator": "some", "value": 42}
+        deserializer = option_deserializer(int_deserializer)
+        value = deserializer(serialized)
+        assert value == Option.some(42)
+
+    @staticmethod
+    def test_option_deserializer_none() -> None:
+        serialized: Json = {"discriminator": "nothing", "value": None}
+        deserializer = option_deserializer(int_deserializer)
+        value = deserializer(serialized)
+        assert value == Option.nothing()
+
+
+class TestSumSerializer:
+    @staticmethod
+    def test_sum_serializer_left() -> None:
+        value = Sum[int, str].left(42)
+        serializer = sum_serializer(int_serializer, str_serializer)
+        serialized = serializer(value)
+        assert serialized == {"discriminator": "left", "value": 42}
+
+    @staticmethod
+    def test_sum_serializer_right() -> None:
+        value = Sum[int, str].right("42")
+        serializer = sum_serializer(int_serializer, str_serializer)
+        serialized = serializer(value)
+        assert serialized == {"discriminator": "right", "value": "42"}
+
+    @staticmethod
+    def test_sum_deserializer_left() -> None:
+        serialized: Json = {"discriminator": "left", "value": 42}
+        deserializer = sum_deserializer(int_deserializer, str_deserializer)
+        value = deserializer(serialized)
+        assert value == Sum[int, str].left(42)
+
+    @staticmethod
+    def test_sum_deserializer_right() -> None:
+        serialized: Json = {"discriminator": "right", "value": "42"}
+        deserializer = sum_deserializer(int_deserializer, str_deserializer)
+        value = deserializer(serialized)
+        assert value == Sum[int, str].right("42")
+
+
+class TestComplexTypesSerializer:
```

Comment:
Keep these tests focused, for each generic type, write one test ensuring that the recursion works, but keep it minimal (no need for complex types in the nested one)

Reply:
I assume you mean nesting, not recursion (there is no recursion here)